### PR TITLE
Start a full API and test suite rewrite

### PR DESCRIPTION
**Pull request by mara004**
Experimental change.

#### Tasks
* Add remaining tests
  * CLI
  * Page Renderer: Annotations, Forms
* Improve global namespace
* Think of better names for some parts
* Extend API (e. g. page labels, page insertion, document metadata, viewer preferences ...) → Perhaps better in separate PRs

#### Done (mostly)
* New API and documentation
* Changelog
* Tests for version, renderer, opener, saver, text extractor, link extractor, text inserter, pdf boxes, and toc
* Setup and version improvements

#### Post-merge tasks
* Update the [team readme](https://github.com/pypdfium2-team/.github/blob/main/profile/README.md) by merging https://github.com/pypdfium2-team/.github/pull/1
* Update the [stackoverflow example](https://stackoverflow.com/questions/46184239/extract-a-page-from-a-pdf-as-a-jpeg/70228315#70228315).
* Submit patches for `doctr`, `py-pdf/benchmarks` and `Extract-URLs` to switch to the new API.

### CODE REVIEW

**Comment by mara004** [comment]
In case someone would like to review or beta-test this change: @MartinThoma @frgfm @adam-huganir @mkroutikov

File: src/pypdfium2/_helpers/_utils.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+import pypdfium2._pypdfium as pdfium
+
+
+def colour_tohex(r, g, b, a=255):
+    """
+    Convert an RGBA colour specified by four integers ranging from 0 to 255 to a single ARGB32 value.
+    
+    Returns:
+        (int, bool) – The colour value, and a boolean signifying if an alpha channel is needed.
```
Comment (mara004) [inline src/pypdfium2/_helpers/_utils.py]: `:` instead of `-`

File: src/pypdfium2/_helpers/page.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+            annotations (bool):
+                Whether to render page annotations.
+            
+            greyscale (bool):
+                Whether to render in greyscale mode (no colours).
+            
+            optimise_mode (OptimiseMode):
+                How to optimise page rendering.
+            
+            crop (typing.Sequence[float]):
```
Comment (mara004) [inline src/pypdfium2/_helpers/page.py]: move up to match argument order

File: tests/test_module.py [hunk 3]
```diff
@@ -2,9 +2,7 @@
 # SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
```
Comment (mara004) [inline tests/test_module.py]: merge into `test_setupscripts.py`

**Comment by mara004** [comment]
See also PR https://github.com/pypdfium2-team/pypdfium2/pull/113, which targets this branch.

File: src/pypdfium2/_cli/toc.py [hunk 4]
```diff
@@ -21,21 +21,7 @@ def attach_parser(subparsers):
...
-                item.view_pos,
-            )
-        )
-
-
 def main(args):
     doc = pdfium.PdfDocument(args.input)
-    print_toc( doc.get_toc() )
+    doc.print_toc( doc.get_toc() )
     doc.close()
```
Comment (mara004) [inline src/pypdfium2/_cli/toc.py]: maybe switch to `pdf` naming here for consistency

File: src/pypdfium2/_helpers/_opener.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+def _is_buffer(maybe_buffer):
+    if all( callable(getattr(maybe_buffer, a, None)) for a in ('seek', 'tell', 'readinto') ):
+        return True
+    else:
+        return False
+
+
+def open_pdf_file(filepath, password=None):
+    filepath = os.path.abspath(filepath)
```
Comment (mara004) [inline src/pypdfium2/_helpers/_opener.py]: Should call `expanduser` before `abspath`

File: src/pypdfium2/_helpers/document.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+        """ FPDF_DOCUMENT: The raw PDFium document object handle. """
+        return self._pdf
+    
+    @classmethod
+    def new(cls):
+        """
+        Create a new, empty document.
+        
+        Returns:
+            PdfDocument:
```
Comment (mara004) [inline src/pypdfium2/_helpers/document.py]: improve return doc

File: src/pypdfium2/_helpers/document.py [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+        
+        Parameters:
+            font_path (str):
+                File path of the font to use.
+            type (int):
+                A constant signifying the type of the given font (:data:`.FPDF_FONT_TYPE1` or :data:`.FPDF_FONT_TRUETYPE`).
+            is_cid (bool):
+                State whether the given font is a CID font or not.
+        Returns:
+            PdfFont:
```
Comment (mara004) [inline src/pypdfium2/_helpers/document.py]: improve return doc

File: src/pypdfium2/_helpers/misc.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    Parameters:
+        level (int):
+            Number of parent items.
+        title (str):
+            String of the bookmark.
+        is_closed (bool):
+            If :data:`True`, child items shall be hidden by default.
+        page_index (int):
+            Zero-based index of the page the bookmark points to.
+        view_mode (``PDFDEST_VIEW_...``):
```
Comment (mara004) [inline src/pypdfium2/_helpers/misc.py]: Note as int and add `PDFDEST_VIEW_...` to the description?

File: src/pypdfium2/_helpers/page.py [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+    def count_objects(self):
+        """
+        Returns:
+            int: The number of page objects on this page.
+        """
+        return pdfium.FPDFPage_CountObjects(self._page)
+    
+    def _get_obj(self, index):
+        raw_obj = pdfium.FPDFPage_GetObject(self._page, index)
+        return PdfPageObject(raw_obj)
```
Comment (mara004) [inline src/pypdfium2/_helpers/page.py]: Merge into `get_objects()`?

File: src/pypdfium2/_helpers/page.py [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+        
+        bitmap = pdfium.FPDFBitmap_CreateEx(width, height, cl_pdfium, None, width*n_colours)
+        if fpdf_colour is not None:
+            pdfium.FPDFBitmap_FillRect(bitmap, 0, 0, width, height, fpdf_colour)
+        
+        render_flags = 0
+        if annotations:
+            render_flags |= pdfium.FPDF_ANNOT
+        if greyscale:
+            render_flags |= pdfium.FPDF_GRAYSCALE
```
Comment (mara004) [inline src/pypdfium2/_helpers/page.py:339]: Add more flags, e. g. to disable anti-aliasing. See https://pdfium.googlesource.com/pdfium/+/15d326e3d3fbbf178fde4c22ec8860f08b54362c/public/fpdfview.h#745

File: src/pypdfium2/_helpers/page.py [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+    
+    
+    def render_topil(self, *args, **kwargs):
+        """
+        *Requires* :mod:`PIL`.
+        
+        Rasterise the page to a PIL image. Parameters are the same as for :meth:`.render_base`.
+        
+        Returns:
+            PIL.Image.Image:
```
Comment (mara004) [inline src/pypdfium2/_helpers/page.py]: improve return doc

File: src/pypdfium2/_helpers/textpage.py [hunk 12]
```diff
@@ -161,7 +195,7 @@ def search(self, text, index=0, match_case=False, match_whole_word=False):
             match_whole_word (bool):
                 If :data:`True`, substring occurrences will be ignored (e. g. `cat` would not match `category`).
         Returns:
-            :class:`.TextSearcher`
+            PdfTextSearcher:
```
Comment (mara004) [inline src/pypdfium2/_helpers/textpage.py]: improve return doc

File: setupsrc/pl_setup/packaging_base.py [hunk 13]
```diff
@@ -56,31 +56,53 @@ def run_cmd(command, cwd, capture=False, **kwargs):
...
+    with open(bindings_file, "w") as file_writer:
         file_writer.write(text)
 
 
-def _get_ver_namespace(ver_file):
-    ver_namespace = {}
-    with open(ver_file, 'r') as fh:
-        exec(fh.read(), ver_namespace)
-    return ver_namespace
+def _get_version_ns():
```
Comment (mara004) [inline setupsrc/pl_setup/packaging_base.py]: make public

**Comment by mara004** [comment]
I think this is ready for merging. The few remaining tasks can be done in main now.

**Comment by frgfm** [comment]
> In case someone would like to review or beta-test this change: @MartinThoma @frgfm @adam-huganir @mkroutikov

Sorry @mara004, I was away for some time and wasn't able to help in time :/

**Comment by mara004** [comment]
No problem :) In the end, this PR got such a chaotic mix of stylistic and functional changes that it might have been a bit too hard to overview for an external reviewer anyway.
In case have any questions or concerns about the new API, please let me know.

**Comment by frgfm** [comment]
Haha, I'm sure you handled it alright!
I'll try to help in the future, feel free to ping me :+1:

**Pull request merged**